File: android-sample/app/src/main/kotlin/com/clevertap/android/nativeui/sample/SlotDemoScreen.kt
package com.clevertap.android.nativeui.sample

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import com.clevertap.android.nativedisplay.bridge.NativeDisplayBridge
import com.clevertap.android.nativedisplay.placement.NativeDisplaySlot
import com.clevertap.android.sdk.CleverTapAPI

// --- Data models ---

data class AppContentItem(
    val id: Int,
    val title: String,
    val subtitle: String,
    val imageUrl: String
)

sealed class SlotDemoItem {
    data class AppContent(val item: AppContentItem) : SlotDemoItem()
    data class SlotPlaceholder(val slotId: String) : SlotDemoItem()
}

// --- Hardcoded content ---

private val appItems = listOf(
    AppContentItem(1, "Morning Yoga Flow", "30 min \u00B7 Beginner friendly", "https://yavuzceliker.github.io/sample-images/image-1.jpg"),
    AppContentItem(2, "Mediterranean Salad", "Quick & healthy lunch recipe", "https://yavuzceliker.github.io/sample-images/image-5.jpg"),
    AppContentItem(3, "Productivity Hacks", "5 tips for focused work", "https://yavuzceliker.github.io/sample-images/image-10.jpg"),
    AppContentItem(4, "Trail Running Guide", "Best routes near you", "https://yavuzceliker.github.io/sample-images/image-15.jpg"),
    AppContentItem(5, "Indoor Plants 101", "Low-maintenance greenery", "https://yavuzceliker.github.io/sample-images/image-20.jpg"),
    AppContentItem(6, "Weekend Getaways", "Top 10 road trip destinations", "https://yavuzceliker.github.io/sample-images/image-25.jpg"),
    AppContentItem(7, "Budget Meal Prep", "Save time and money", "https://yavuzceliker.github.io/sample-images/image-30.jpg"),
    AppContentItem(8, "Home Workout", "No equipment needed", "https://yavuzceliker.github.io/sample-images/image-35.jpg"),
    AppContentItem(9, "Coffee Brewing", "Perfect pour-over technique", "https://yavuzceliker.github.io/sample-images/image-40.jpg"),
    AppContentItem(10, "Sleep Better", "Science-backed tips", "https://yavuzceliker.github.io/sample-images/image-45.jpg"),
    AppContentItem(11, "Digital Detox", "Unplug and recharge", "https://yavuzceliker.github.io/sample-images/image-50.jpg"),
    AppContentItem(12, "Book Club Picks", "This month's top reads", "https://yavuzceliker.github.io/sample-images/image-55.jpg"),
    AppContentItem(13, "Smoothie Recipes", "Fuel your morning", "https://yavuzceliker.github.io/sample-images/image-60.jpg"),
    AppContentItem(14, "Desk Stretches", "Relieve tension in 5 min", "https://yavuzceliker.github.io/sample-images/image-65.jpg"),
    AppContentItem(15, "Mindful Breathing", "Calm in 3 minutes", "https://yavuzceliker.github.io/sample-images/image-70.jpg"),
)

/**
 * Builds the interleaved list: 4 slot placeholders at fixed positions among 15 app items.
 *
 * Index mapping:
 *   0          -> slot_top
 *   1-3        -> app items 1-3
 *   4          -> slot_feed_1
 *   5-7        -> app items 4-6
 *   8          -> slot_feed_2
 *   9-17       -> app items 7-15
 *   18         -> slot_bottom
 */
private fun buildFeedItems(): List<SlotDemoItem> {
    val items = mutableListOf<SlotDemoItem>()

    items.add(SlotDemoItem.SlotPlaceholder("slot_top"))          // 0
    items.addAll(appItems.subList(0, 3).map { SlotDemoItem.AppContent(it) })  // 1-3
    items.add(SlotDemoItem.SlotPlaceholder("slot_feed_1"))       // 4
    items.addAll(appItems.subList(3, 6).map { SlotDemoItem.AppContent(it) })  // 5-7
    items.add(SlotDemoItem.SlotPlaceholder("slot_feed_2"))       // 8
    items.addAll(appItems.subList(6, 15).map { SlotDemoItem.AppContent(it) }) // 9-17
    items.add(SlotDemoItem.SlotPlaceholder("slot_bottom"))       // 18

    return items
}

// --- Screen ---

@Composable
fun SlotDemoScreen() {
    val context = LocalContext.current

    // Initialize bridge (same pattern as CleverTapIntegrationScreen)
    val bridge = remember { NativeDisplayBridge.initialize(context.applicationContext) }
    val cleverTapApi = remember { CleverTapAPI.getDefaultInstance(context.applicationContext) }

    val feedItems = remember { buildFeedItems() }

    LazyColumn(
        modifier = Modifier
            .fillMaxSize()
            .background(Color(0xFFF5F5F5)),
        contentPadding = PaddingValues(horizontal = 16.dp, vertical = 12.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        // Header section
        item {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = 4.dp)
            ) {
                Text(
                    text = "Slot Demo",
                    style = MaterialTheme.typography.headlineSmall,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier.padding(bottom = 4.dp)
                )
                Text(
                    text = "This feed contains 4 NativeDisplaySlot views at fixed positions. " +
                            "Tap the button below to fire a CleverTap event that fetches real server data for the slots.",
                    style = MaterialTheme.typography.bodyMedium,
                    color = Color(0xFF666666),
                    modifier = Modifier.padding(bottom = 12.dp)
                )
                Button(
                    onClick = {
                        cleverTapApi?.run {
                            pushEvent("Header1")
                            pushEvent("Header2")
                            pushEvent("Header3")
                            pushEvent("lalit")
                        }
                    },
                    modifier = Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(8.dp)
                ) {
                    Text("Fetch Slot Data")
                }
            }
        }

        // Feed items (19 total: 4 slots + 15 app content)
        itemsIndexed(feedItems) { _, item ->
            when (item) {
                is SlotDemoItem.SlotPlaceholder -> {
                    NativeDisplaySlot(
                        slotId = item.slotId,
                        modifier = Modifier.fillMaxWidth(),
                        loading = { EmptySlotPlaceholder() }
                    )
                }
                is SlotDemoItem.AppContent -> {
                    AppContentCard(item.item)
                }
            }
        }
    }
}

// --- Slot placeholder (shown before data arrives) ---

@Composable
private fun EmptySlotPlaceholder() {
    val dashedBorderColor = Color(0xFFBDBDBD)

    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(80.dp)
            .background(
                color = Color(0xFFF5F5F5),
                shape = RoundedCornerShape(8.dp)
            )
            .drawBehind {
                val strokeWidth = 1.dp.toPx()
                val pathEffect = PathEffect.dashPathEffect(
                    floatArrayOf(8.dp.toPx(), 4.dp.toPx()),
                    0f
                )
                drawRoundRect(
                    color = dashedBorderColor,
                    style = Stroke(
                        width = strokeWidth,
                        pathEffect = pathEffect
                    ),
                    cornerRadius = CornerRadius(8.dp.toPx())
                )
            },
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = "Ad",
            color = Color(0xFF9E9E9E),
            fontSize = 14.sp,
            fontWeight = FontWeight.Medium
        )
    }
}

// --- App content card ---

@Composable
private fun AppContentCard(item: AppContentItem) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(12.dp),
        colors = CardDefaults.cardColors(containerColor = Color.White),
        elevation = CardDefaults.cardElevation(defaultElevation = 2.dp)
    ) {
        Column {
            AsyncImage(
                model = item.imageUrl,
                contentDescription = item.title,
                contentScale = ContentScale.Crop,
                modifier = Modifier
                    .fillMaxWidth()
                    .height(180.dp)
                    .clip(RoundedCornerShape(topStart = 12.dp, topEnd = 12.dp))
            )
            Column(modifier = Modifier.padding(12.dp)) {
                Text(
                    text = item.title,
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis
                )
                Spacer(modifier = Modifier.height(4.dp))
                Text(
                    text = item.subtitle,
                    style = MaterialTheme.typography.bodyMedium,
                    color = Color(0xFF888888),
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis
                )
            }
        }
    }
}

File: flutter-sample/lib/screens/slot_demo_screen.dart
import 'package:flutter/material.dart';

// --- Data models ---

class _AppItem {
  final int id;
  final String title;
  final String subtitle;
  final String imageUrl;
  const _AppItem(this.id, this.title, this.subtitle, this.imageUrl);
}

sealed class _FeedItem {}
class _ContentItem extends _FeedItem {
  final _AppItem item;
  _ContentItem(this.item);
}
class _SlotItem extends _FeedItem {
  final String slotId;
  _SlotItem(this.slotId);
}

// --- Feed data ---

const _appItems = [
  _AppItem(1,  'Morning Yoga Flow',      '30 min · Beginner friendly',         'https://yavuzceliker.github.io/sample-images/image-1.jpg'),
  _AppItem(2,  'Mediterranean Salad',    'Quick & healthy lunch recipe',        'https://yavuzceliker.github.io/sample-images/image-5.jpg'),
  _AppItem(3,  'Productivity Hacks',     '5 tips for focused work',             'https://yavuzceliker.github.io/sample-images/image-10.jpg'),
  _AppItem(4,  'Trail Running Guide',    'Best routes near you',                'https://yavuzceliker.github.io/sample-images/image-15.jpg'),
  _AppItem(5,  'Indoor Plants 101',      'Low-maintenance greenery',            'https://yavuzceliker.github.io/sample-images/image-20.jpg'),
  _AppItem(6,  'Weekend Getaways',       'Top 10 road trip destinations',       'https://yavuzceliker.github.io/sample-images/image-25.jpg'),
  _AppItem(7,  'Budget Meal Prep',       'Save time and money',                 'https://yavuzceliker.github.io/sample-images/image-30.jpg'),
  _AppItem(8,  'Home Workout',           'No equipment needed',                 'https://yavuzceliker.github.io/sample-images/image-35.jpg'),
  _AppItem(9,  'Coffee Brewing',         'Perfect pour-over technique',         'https://yavuzceliker.github.io/sample-images/image-40.jpg'),
  _AppItem(10, 'Sleep Better',           'Science-backed tips',                 'https://yavuzceliker.github.io/sample-images/image-45.jpg'),
  _AppItem(11, 'Digital Detox',          'Unplug and recharge',                 'https://yavuzceliker.github.io/sample-images/image-50.jpg'),
  _AppItem(12, 'Book Club Picks',        'This month\'s top reads',             'https://yavuzceliker.github.io/sample-images/image-55.jpg'),
  _AppItem(13, 'Smoothie Recipes',       'Fuel your morning',                   'https://yavuzceliker.github.io/sample-images/image-60.jpg'),
  _AppItem(14, 'Desk Stretches',         'Relieve tension in 5 min',            'https://yavuzceliker.github.io/sample-images/image-65.jpg'),
  _AppItem(15, 'Mindful Breathing',      'Calm in 3 minutes',                   'https://yavuzceliker.github.io/sample-images/image-70.jpg'),
];

List<_FeedItem> _buildFeed() => [
  _SlotItem('slot_top'),
  ..._appItems.sublist(0, 3).map(_ContentItem.new),
  _SlotItem('slot_feed_1'),
  ..._appItems.sublist(3, 6).map(_ContentItem.new),
  _SlotItem('slot_feed_2'),
  ..._appItems.sublist(6, 15).map(_ContentItem.new),
  _SlotItem('slot_bottom'),
];

// --- Screen ---

class SlotDemoScreen extends StatefulWidget {
  const SlotDemoScreen({super.key});

  @override
  State<SlotDemoScreen> createState() => _SlotDemoScreenState();
}

class _SlotDemoScreenState extends State<SlotDemoScreen> {
  final _feed = _buildFeed();
  // In a real integration, slot configs would come from NativeDisplayBridge.
  // Here we show dashed placeholders to demonstrate the layout.

  void _fetchSlotData() {
    ScaffoldMessenger.of(context).showSnackBar(
      const SnackBar(
        content: Text('Integrate CleverTap Core SDK to receive live slot data'),
        duration: Duration(seconds: 3),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: const Color(0xFFF5F5F5),
      appBar: AppBar(
        title: const Text('Slot Demo'),
        backgroundColor: Theme.of(context).colorScheme.primary,
        foregroundColor: Theme.of(context).colorScheme.onPrimary,
      ),
      body: ListView.builder(
        padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 12),
        itemCount: _feed.length + 1, // +1 for header
        itemBuilder: (ctx, i) {
          if (i == 0) return _Header(onFetch: _fetchSlotData);
          final item = _feed[i - 1];
          return Padding(
            padding: const EdgeInsets.only(bottom: 12),
            child: switch (item) {
              _SlotItem s => _SlotPlaceholder(slotId: s.slotId),
              _ContentItem c => _AppContentCard(item: c.item),
            },
          );
        },
      ),
    );
  }
}

// --- Header ---

class _Header extends StatelessWidget {
  final VoidCallback onFetch;
  const _Header({required this.onFetch});

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.only(bottom: 12),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          const Text(
            'Slot Demo',
            style: TextStyle(fontSize: 22, fontWeight: FontWeight.bold),
          ),
          const SizedBox(height: 4),
          const Text(
            'This feed contains 4 NativeDisplaySlot views at fixed positions. '
            'Tap the button below to fire a CleverTap event that fetches real server data for the slots.',
            style: TextStyle(fontSize: 14, color: Color(0xFF666666)),
          ),
          const SizedBox(height: 12),
          SizedBox(
            width: double.infinity,
            child: FilledButton(
              onPressed: onFetch,
              style: FilledButton.styleFrom(
                shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
              ),
              child: const Text('Fetch Slot Data'),
            ),
          ),
          const SizedBox(height: 4),
        ],
      ),
    );
  }
}

// --- Slot placeholder (dashed border, matches Android EmptySlotPlaceholder) ---

class _SlotPlaceholder extends StatelessWidget {
  final String slotId;
  const _SlotPlaceholder({required this.slotId});

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      painter: _DashedBorderPainter(),
      child: SizedBox(
        height: 80,
        width: double.infinity,
        child: Center(
          child: Text(
            'Ad',
            style: TextStyle(
              color: Colors.grey[500],
              fontSize: 14,
              fontWeight: FontWeight.w500,
            ),
          ),
        ),
      ),
    );
  }
}

class _DashedBorderPainter extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    const radius = Radius.circular(8);
    final rect = RRect.fromLTRBR(0, 0, size.width, size.height, radius);
    final paint = Paint()
      ..color = const Color(0xFFBDBDBD)
      ..style = PaintingStyle.stroke
      ..strokeWidth = 1;

    const dashLen = 8.0;
    const gapLen = 4.0;
    final path = Path()..addRRect(rect);
    final pathMetrics = path.computeMetrics();
    for (final metric in pathMetrics) {
      var distance = 0.0;
      while (distance < metric.length) {
        final end = (distance + dashLen).clamp(0.0, metric.length);
        canvas.drawPath(metric.extractPath(distance, end), paint);
        distance += dashLen + gapLen;
      }
    }
  }

  @override
  bool shouldRepaint(_DashedBorderPainter old) => false;
}

// --- App content card ---

class _AppContentCard extends StatelessWidget {
  final _AppItem item;
  const _AppContentCard({required this.item});

  @override
  Widget build(BuildContext context) {
    return Card(
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12)),
      color: Colors.white,
      elevation: 2,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          ClipRRect(
            borderRadius: const BorderRadius.vertical(top: Radius.circular(12)),
            child: Image.network(
              item.imageUrl,
              height: 180,
              width: double.infinity,
              fit: BoxFit.cover,
              errorBuilder: (_, __, ___) => Container(
                height: 180,
                color: const Color(0xFFEEEEEE),
                child: const Icon(Icons.image, color: Color(0xFFBBBBBB), size: 48),
              ),
              loadingBuilder: (_, child, progress) => progress == null
                  ? child
                  : Container(
                      height: 180,
                      color: const Color(0xFFF5F5F5),
                      child: const Center(child: CircularProgressIndicator(strokeWidth: 2)),
                    ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(12),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  item.title,
                  style: const TextStyle(fontWeight: FontWeight.w600, fontSize: 15),
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                ),
                const SizedBox(height: 4),
                Text(
                  item.subtitle,
                  style: const TextStyle(color: Color(0xFF888888), fontSize: 13),
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}
